### ISSUE TITLE
`TextBlock` Foreground binding doesn't cause re-render in some scenarios

### ISSUE THREAD

**Issue by SKProCH** [AvaloniaUI/Avalonia#9675]
**Describe the bug**
`TextBlock` doesn't re-render when Foreground brush color are changed in some scenarios (details below). It looks like that:
![182f3c5d-c9e6-46cf-afe3-606df8a7cf0c](https://user-images.githubusercontent.com/29896317/206933822-ca864faf-531b-432d-8b7e-67aeed29a2ac.gif)
My guess is that this happens when a `TextBlock` is inside an element without the `TextElement.Foreground` property (i.e. inside a control that is not inherited from a `TemplatedControl`, like `StackPanel`, `Border`, etc), and they are together inside the template (`ControlTemplate`) of some control.

**To Reproduce**
Steps to reproduce the behavior (link to repository with repro at the end):

1. Define a new style in App.Styles to contain our SolidColorBrush (it can be in App.Resources directly, but nothing changes)
    ```xaml
    <Style>
      <Style.Resources>
        <SolidColorBrush x:Key="MaterialDesignBody" Color="Aqua"></SolidColorBrush>
      </Style.Resources>
    </Style>
    ```
2. Add a new control, inherited from `TemplatedControl` and set its template:
    ```csharp
    public class MyTestControl : TemplatedControl { }
    ```
    ```xaml
    <ControlTheme x:Key="{x:Type local:MyTestControl}" TargetType="local:MyTestControl">
      <Setter Property="Template">
        <ControlTemplate>
          <StackPanel> <!-- StackPanel IS IMPORTANT -->
            <TextBlock>BROKEN COLORED TEXT TEXTBLOCK</TextBlock>
          </StackPanel> <!-- StackPanel IS IMPORTANT -->
        </ControlTemplate>
      </Setter>
    </ControlTheme>
    ```
3. Set `TextElement.Foreground` to root element (MainView for example) to set color for all text elements in this element:
    ```xaml
    <UserControl ...
                 TextElement.Foreground="{DynamicResource MaterialDesignBody}">
    ```
4. Add simple `TextBlock` (to see difference) and our `MyTestControl` to MainView (or elsewhere):
    ```xaml
    <TextBlock>NORMAL COLORED TEXT TEXTBLOCK</TextBlock>
    <avaloniaChangeColorRepro:MyTestControl />
    ```
5. Add color changing code (bind it to button in my case):
    ```csharp
    private bool _isDark;
    private void Button_OnClick(object? sender, RoutedEventArgs e) {
        _isDark = !_isDark;
        var c = _isDark ? Color.Parse("White") : Color.Parse("Black");
        var style = App.Current.Styles.Last() as Style;
        var resource = style.Resources["MaterialDesignBody"] as SolidColorBrush;
        resource.Color = c;
    }
    ```
6. See result (gif above)

**Expected behavior**
All `TextBlock` must re-rendered when color in brush was changed.

**Desktop (please complete the following information):**

- Version: `11.0.0-preview4`

**Additional context**
If we remove and add resource again, all TextBlock will update and re-render properly:
![22e04f4c-4fbd-4199-a021-b68061480b3f](https://user-images.githubusercontent.com/29896317/206934610-d4a3f0fe-ba51-4efe-ac94-c2545a9bc5ce.gif)
```csharp
private void ButtonReAdd_OnClick(object? sender, RoutedEventArgs e) {
    var style = App.Current.Styles.Last() as Style;
    App.Current.Styles.Remove(style);
    App.Current.Styles.Add(style);
}
```
But this only works for `TextBlock`s that were already displayed at the time the resource was re-added. If we add new `TextBlock`es (for example, by switching to another `UserControl`), the problem persists only for them (the new elements).

Full repro for this can be found here: https://github.com/SKProCH/AvaloniaChangeColorRepro
Material.Avalonia issue about that can be found here: https://github.com/AvaloniaCommunity/Material.Avalonia/issues/200 (contains additional details and workarounds)
This problem makes runtime Material.Avalonia's theme changing broken.

**Comment by robloo** [comment]
I've seen this one for a long time. There have been various fixes different areas and I somewhat expect https://github.com/AvaloniaUI/Avalonia/pull/9631 will resolve it. Can you test with that PR and see if it changes things?

**Comment by SKProCH** [comment]
Seems like it fixed with preview-5. Need more comprehensive testing.

**Comment by SKProCH** [comment]
Yep, everything working fine

**Comment by bmscodespace** [comment]
Hi,

I am using Avalonia 0.10.21 and Material Avalonia 2.5.1 and I am having a similar problem with the "Background" binding on several controls:
While something like the source binding f.e. on 
` <Image  Source="{DynamicResource SomeImage}" />`
works the following doesn't:
`<ProgressBar ... Background="{DynamicResource SomeColor}" />`
Is this fixed in 0.10.22?

**Comment by maxkatz6** [comment]
@bmscodespace no, this specific problem was only introduced in 11.0 previews and fixed in the next preview version.
